deepinv/optim/distance.py

```diff
@@ -0,0 +1,364 @@
+        """
+        return (x + gamma * y) / (1 + gamma)
+
+
+class IndicatorL2Distance(Distance):
+    r"""
+    Indicator of :math:`\ell_2` ball with radius :math:`r`.
+
+    The indicator function of the $\ell_2$ ball with radius :math:`r`, denoted as \iota_{\mathcal{B}_2(y,r)(x)},
+    is defined as
+
+    .. math::
+
+          \iota_{\mathcal{B}_2(y,r)}(x)= \left.
+              \begin{cases}
+                0, & \text{if } \|x-y\|_2\leq r \\
+                +\infty & \text{else.}
+              \end{cases}
+              \right.
+
+
+    :param float radius: radius of the ball. Default: None.
+
+    """
+
+    def __init__(self, radius=None):
+        super().__init__()
+        self.radius = radius
+
+    def fn(self, x, y, *args, radius=None, **kwargs):
+        r"""
+        Computes the batched indicator of :math:`\ell_2` ball with radius `radius`, i.e. :math:`\iota_{\mathcal{B}(y,r)}(x)`.
+
+        :param torch.Tensor x: Variable :math:`x` at which the indicator is computed. :math:`u` is assumed to be of shape (B, ...) where B is the batch size.
+        :param torch.Tensor y: Observation :math:`y` of the same dimension as :math:`u`.
+        :param float radius: radius of the :math:`\ell_2` ball. If `radius` is None, the radius of the ball is set to `self.radius`. Default: None.
+        :return: (torch.Tensor) indicator of :math:`\ell_2` ball with radius `radius`. If the point is inside the ball, the output is 0, else it is 1e16.
+        """
+        diff = x - y
+        dist = torch.norm(diff.reshape(diff.shape[0], -1), p=2, dim=-1)
+        radius = self.radius if radius is None else radius
+        loss = (dist > radius) * 1e16
+        return loss
+
+    def prox(self, x, y, *args, radius=None, gamma=None, **kwargs):
+        r"""
+        Proximal operator of the indicator of :math:`\ell_2` ball with radius `radius`, i.e.
+
+        .. math::
+
+            \operatorname{prox}_{\iota_{\mathcal{B}_2(y,r)}}(x) = \operatorname{proj}_{\mathcal{B}_2(y, r)}(x)
+
+
+        where :math:`\operatorname{proj}_{C}(x)` denotes the projection on the closed convex set :math:`C`.
+
+
+        :param torch.Tensor x: Variable :math:`x` at which the proximity operator is computed.
+        :param torch.Tensor y: Observation :math:`y` of the same dimension as :math:`x`.
+        :param float gamma: step-size. Note that this parameter is not used in this function.
+        :param float radius: radius of the :math:`\ell_2` ball.
+        :return: (torch.Tensor) projection on the :math:`\ell_2` ball of radius `radius` and centered in `y`.
+        """
+        radius = self.radius if radius is None else radius
+        diff = x - y
+        dist = torch.norm(diff.reshape(diff.shape[0], -1), p=2, dim=-1)
+        return y + diff * (
+            torch.min(torch.tensor([radius]).to(x.device), dist) / (dist + 1e-12)
+        ).view(-1, 1, 1, 1)
+
+
+class KullbackLeiblerDistance(Distance):
```

Comment:
I would keep this name to 'poisson log likelihood' since the KL divergence is not specific to the Poisson distribution.